### PR TITLE
Break up counter end to end test beforeEach hook for more granular error info

### PR DESCRIPTION
**Pull request by ChumpChief**
Counter increment/decrement has been flagged as a flaky test.  Nothing seems to indicate it is a counter-specific issue, but it's a reasonable one to try and diagnose further (relatively simple feature and test).

This PR just breaks up the beforeEach hook into multiple stages which will show up in failure logs to better understand where the flakiness occurs.  Currently we just see that "setup" is timing out, but with this PR we should see more specifically what part of that setup.

Locally, testing with t9s (one of the failing environments, see run 271404) I actually get *mostly* reliable failures.  Specifically I get almost-consistent failure with 1.4.0 LOADER ONLY to either (1) load container2 & container3 or (2) synchronize the "increment 7" counter op across clients (times out with the sending client still dirty).  I want to merge this change to see how similar the results are both for t9s but also for other services too.

[AB#8335](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8335)

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/counterEndToEndTests.spec.ts [hunk 1]
```diff
@@ -37,27 +37,46 @@ describeCompat("SharedCounter", "FullCompat", (getTestObjectProvider, apis) => {
...
+	let dataStore3: ITestFluidObject;
 	let sharedCounter1: ISharedCounter;
 	let sharedCounter2: ISharedCounter;
 	let sharedCounter3: ISharedCounter;
 
-	beforeEach("setup", async () => {
-		// Create a Container for the first client.
-		const container1 = await provider.makeTestContainer(testContainerConfig);
+	// To help narrow down test flakiness, the beforeEach is broken into tiny phases.  These phase names will
+	// show up in error logs to help identify where timeouts are occurring.
```
Comment (alexvy86) [inline packages/test/test-end-to-end-tests/src/test/counterEndToEndTests.spec.ts:51]:
```suggestion
	// To help narrow down test flakiness, the beforeEach is broken into tiny phases.  These phase names will
	// show up in error logs to help identify where timeouts are occurring.
	// Note that their execution order matters. As of 2024-08-20, Mocha runs them in the order they are defined.
```

**Review by alexvy86 — APPROVED** [review]
Nothing against the refactor, I'll just note that it already changes the order of awaits compared to before, so it might have an impact on the timeout's reproducibility.

I'll also mention that debugging e2e tests against local server, one thing @markfields and I ran into is that some tests that want to wait for a summary do it "too late". They'll make changes to a DDS, then do async stuff, and only afterwards set up listeners for the summary ops/acks. Which many times seems to work fine, but could actually miss the op/ack if the summary happened quickly enough, during any other awaits, before listeners were set up; and in that case the tests time out. In case you run into a pattern like that.

File: packages/test/test-end-to-end-tests/src/test/counterEndToEndTests.spec.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/test/test-end-to-end-tests/src/test/counterEndToEndTests.spec.ts:51]: Added a similar note in latest

**Comment by ChumpChief** [comment]
> Nothing against the refactor, I'll just note that it already changes the order of awaits compared to before, so it might have an impact on the timeout's reproducibility.

I think you are referring to e.g. how now all the containers load before the first data object is retrieved, etc.?  Agree this will affect the test run, but I suspect it will not change the root cause.  In particular, the failures of type (2) are spooky, that it appears tinylicious is not ack'ing the op (or somehow the sender is failing to actually send it).  These hit even with a 20s timeout so I think there's probably something functional to track down and the await ordering won't matter for that.

I can observe what appear to be similar failures in other DDS tests (e.g. for map) so we can also use those as a control group if/when we diagnose the issue with this test.

**Pull request merged**